pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,604 @@
+	return vec.downstreamVector.GetNulls() != nil && vec.downstreamVector.GetNulls().Contains(uint64(i))
+}
+
+func (vec *CnTaeVector[T]) NullMask() *roaring64.Bitmap {
+	if input := vec.downstreamVector.GetNulls().Np; input != nil {
+		np := roaring64.New()
+		np.AddMany(input.ToArray())
+		return np
+	}
+	return nil
+}
+
+func (vec *CnTaeVector[T]) GetType() types.Type {
+	return vec.downstreamVector.GetType()
+}
+
+func (vec *CnTaeVector[T]) Compact(deletes *roaring.Bitmap) {
+	var sels []int64
+	vecLen := uint32(vec.Length())
+	for i := uint32(0); i < vecLen; i++ {
+		if !deletes.Contains(i) {
+			sels = append(sels, int64(i))
+		}
+	}
+	cnVector.Shrink(vec.downstreamVector, sels)
+}
+
+func (vec *CnTaeVector[T]) String() string {
+	s := fmt.Sprintf("DN Vector: Len=%d[Rows];Cap=%d[Rows];Allocted:%d[Bytes]", vec.Length(), vec.Capacity(), vec.Allocated())
+
+	end := 100
+	if vec.Length() < end {
+		end = vec.Length()
+	}
+	if end == 0 {
+		return s
+	}
+
+	data := "Vals=["
+	for i := 0; i < end; i++ {
+		data = fmt.Sprintf("%s %v", data, vec.Get(i))
+	}
+	if vec.Length() > end {
+		s = fmt.Sprintf("%s %s...]", s, data)
+	} else {
+		s = fmt.Sprintf("%s %s]", s, data)
+	}
+
+	return s
+}
+
+func (vec *CnTaeVector[T]) Extend(src Vector) {
+	vec.ExtendWithOffset(src, 0, src.Length())
+}
+
+func (vec *CnTaeVector[T]) ExtendWithOffset(src Vector, srcOff, srcLen int) {
+
+	if srcLen <= 0 {
+		return
+	}
+
+	// The downstream vector, ie CN vector needs isNull as argument.
+	// So, we can't directly call cn_vector.Append() without parsing the data.
+	// Hence, we are using src.Get(i) to retrieve the Null value as such from the src, and inserting
+	// it into the current CnVectorAdapter via ExtendWithOffset().
+	for i := srcOff; i < srcOff+srcLen; i++ {
+		vec.Append(src.Get(i))
+	}
+}
+
+func (vec *CnTaeVector[T]) Update(i int, v any) {
+	UpdateValue(vec.downstreamVector, uint32(i), v)
+}
+
+func (vec *CnTaeVector[T]) Reset() {
+	if vec.Length() == 0 {
+		return
+	}
+
+	if vec.Nullable() {
+		cnNulls.Reset(vec.downstreamVector.Nsp)
+		//NOTE: We are not resetting the isNullable.
+	}
+
+	cnVector.Reset(vec.downstreamVector)
+	vec.isAllocatedFromMpool = false
+}
+
+func (vec *CnTaeVector[T]) Slice() any {
+	return vec.downstreamVector.Col
+}
+
+func (vec *CnTaeVector[T]) Bytes() *Bytes {
+	return MoVecToBytes(vec.downstreamVector)
+}
+
+func (vec *CnTaeVector[T]) Foreach(op ItOp, sels *roaring.Bitmap) error {
+	return vec.ForeachWindow(0, vec.Length(), op, sels)
+}
+
+func (vec *CnTaeVector[T]) WriteTo(w io.Writer) (n int64, err error) {
+	var nr int
+
+	// 1. Nullable Flag
+	if nr, err = w.Write(types.EncodeFixed(vec.Nullable())); err != nil {
+		return
+	}
+	n += int64(nr)
+
+	// 2. DownStream Vector
+	output, err := vec.downstreamVector.MarshalBinary()
+	if nr, err = w.Write(output); err != nil {
+		return
+	}
+	n += int64(nr)
+
+	return
+}
+
+func (vec *CnTaeVector[T]) ReadFrom(r io.Reader) (n int64, err error) {
+	// Nullable Flag [1 byte]
```

Comment:
Do you really need all these make?   How about make one and reuse?

Reply:
Sure. I will look into it.


Since 
- we are accepting `io.Reader` instead of []byte
- and the `io.Reader's` output `[]byte` contains variable length

I am not sure how to use only one `[]byte`, to read the whole content.

I was following a similar approach from the [original DN vector](https://github.com/arjunsk/matrixone/blob/e5e451c6bee40b5c116bdd5e699dadb29d07059d/pkg/vm/engine/tae/containers/vector.go#L261) ReadFrom(..)

Reply:
I tried using the io.ReadAll() as shown in the below code

```go
func (vec *CnTaeVector[T]) WriteTo(w io.Writer) (n int64, err error) {

	var buff []byte

	// 1. Nullable Flag
	buff = append(buff, types.EncodeFixed(vec.Nullable())...)

	// 2. DownStream Vector
	var vecBuff []byte
	if vecBuff, err = vec.downstreamVector.MarshalBinary(); err != nil {
		return 0, err
	} else {
		buff = append(buff, vecBuff...)
	}

	if _, err = w.Write(buff); err != nil {
		return 0, err
	}
	n = int64(len(buff))

	return
}

func (vec *CnTaeVector[T]) ReadFrom(r io.Reader) (n int64, err error) {

	var buf []byte
	if buf, err = io.ReadAll(r); err != nil {
		return 0, err
	}

	// Nullable Flag [1 byte]
	nullable := types.DecodeFixed[bool](buf[:1])
	vec.isNullable = nullable

	buf = buf[1:]

	// DownStream Vector
	vec.releaseDownstream()
	vec.downstreamVector = cnVector.New(vec.GetType())

	if err = vec.downstreamVector.UnmarshalBinary(buf); err != nil {
		return 0, err
	}
	n = int64(len(buf))
	return 0, err
}
```

But, it is failing in [TestBatch1(..)](https://github.com/matrixorigin/matrixone/blob/fbe23c4b6e0f352d7b23976f3afcabf1167ebd59/pkg/vm/engine/tae/containers/batch_test.go#L56).

Reason:
In `ReadFrom(..)`, I am getting `len(buf) = 0` for one `batch.Vector`. Is it recommended to use `io.ReadAll(..)` ?

Reply:
Solved using

```go
func (vec *vector[T]) WriteTo(w io.Writer) (n int64, err error) {
	// 1. Nullable Flag [1 byte]
	buf := types.EncodeFixed(vec.Nullable())

	// 2. Vector bytes
	var vecBytes []byte
	if vecBytes, err = vec.downstreamVector.MarshalBinary(); err != nil {
		return 0, err
	}
	buf = append(buf, vecBytes...)

	//0. Length of vector bytes [8 bytes]
	i64 := int64(len(buf))
	buf = append(types.EncodeInt64(&i64), buf...)

	var writtenBytes int
	if writtenBytes, err = w.Write(buf); err != nil {
		return 0, err
	}

	return int64(writtenBytes), nil
}

func (vec *vector[T]) ReadFrom(r io.Reader) (n int64, err error) {
	// 0. Length [8 bytes]
	lengthBytes := make([]byte, 8)
	if _, err = r.Read(lengthBytes); err != nil {
		return 0, err
	}
	length := types.DecodeInt64(lengthBytes[:8])
	n += 8

	// Whole DN Vector
	buf := make([]byte, length)
	if _, err = r.Read(buf); err != nil {
		return 0, err
	}

	n += int64(len(buf))

	// 1. Nullable flag [1 byte]
	isNullable := types.DecodeFixed[bool](buf[:1])
	vec.isNullable = isNullable

	//2. Vector
	buf = buf[1:]

	vec.releaseDownstream()
	if err = vec.downstreamVector.UnmarshalBinary(buf); err != nil {
		return 0, err
	}

	return n, nil
}
```